File: code/traditional/EasyTL/demo_amazon_review.m
%% Test Amazon review dataset

addpath('data/text/amazon_review_400/');
str_domains = {'books','dvd','elec','kitchen'};
list_acc = [];
for i = 1 : 4
    for j = 1 : 4
        if i == j
            continue;
        end
        fprintf('%s - %s\n',str_domains{i}, str_domains{j});
        
        %% Load data
        load([str_domains{i},'_400.mat']);
        Xs = fts;    clear fts;
        Ys = labels; clear labels;
        load([str_domains{j},'_400.mat']);
        Xt = fts;    clear fts;
        Yt = labels; clear labels;
        Ys = Ys + 1;
        Yt = Yt + 1;
        
        Xs = Xs ./ repmat(sum(Xs,2),1,size(Xs,2)); 
        Xs = zscore(Xs,1);
        Xt = Xt ./ repmat(sum(Xt,2),1,size(Xt,2)); 
        Xt = zscore(Xt,1);

        % EasyTL without intra-domain alignment [EasyTL(c)]
        [Acc1, ~] = EasyTL(Xs,Ys,Xt,Yt,'raw');
        fprintf('Acc: %f\n',Acc1);
        
        % EasyTL with CORAL for intra-domain alignment
        [Acc2, ~] = EasyTL(Xs,Ys,Xt,Yt);
        fprintf('Acc: %f\n',Acc2);
        
        list_acc = [list_acc;[Acc1,Acc2]];
        
    end
end

File: code/traditional/pyEasyTL/demo_amazon_review.py
import scipy.io
import scipy.stats
import numpy as np
from EasyTL import EasyTL
import time

if __name__ == "__main__":
	datadir = r"D:\Datasets\EasyTL\amazon_review"
	str_domain = ["books", "dvd", "elec", "kitchen"]
	list_acc = []
	
	for i in range(len(str_domain)):
		for j in range(len(str_domain)):
			if i == j:
				continue
			
			print("{} - {}".format(str_domain[i], str_domain[j]))
			
			mat1 = scipy.io.loadmat(datadir + "/{}_400.mat".format(str_domain[i]))
			Xs = mat1["fts"]
			Ys = mat1["labels"]
			mat2 = scipy.io.loadmat(datadir + "/{}_400.mat".format(str_domain[j]))
			Xt = mat2["fts"]
			Yt = mat2["labels"]
			
			Ys += 1
			Yt += 1
            
			Xs = Xs / np.tile(np.sum(Xs,axis=1).reshape(-1,1), [1, Xs.shape[1]])
			Xs = scipy.stats.mstats.zscore(Xs);
			Xt = Xt / np.tile(np.sum(Xt,axis=1).reshape(-1,1), [1, Xt.shape[1]])
			Xt = scipy.stats.mstats.zscore(Xt);
			
			Xs[np.isnan(Xs)] = 0
			Xt[np.isnan(Xt)] = 0
            
			t0 = time.time()
			Acc1, _ = EasyTL(Xs,Ys,Xt,Yt,"raw")
			t1 = time.time()
			print("Time Elapsed: {:.2f} sec".format(t1 - t0))
			Acc2, _ = EasyTL(Xs,Ys,Xt,Yt)
			t2 = time.time()
			print("Time Elapsed: {:.2f} sec".format(t2 - t1))
            
			print('EasyTL(c) Acc: {:.1f} % || EasyTL Acc: {:.1f} %'.format(Acc1*100, Acc2*100))
			list_acc.append([Acc1,Acc2])
            
	acc = np.array(list_acc)
	avg = np.mean(acc, axis=0)
	print('EasyTL(c) AVG Acc: {:.1f} %'.format(avg[0]*100))
	print('EasyTL AVG Acc: {:.1f} %'.format(avg[1]*100))